### PR TITLE
Warp song particles

### PR DESCRIPTION
**Pull request by PurpleHato**
Fixes: https://github.com/briaguya-ai/rando-issue-tracker/issues/43 , https://github.com/briaguya-ai/rando-issue-tracker/issues/18 , https://github.com/briaguya-ai/rando-issue-tracker/issues/6

### CODE REVIEW

File: soh/src/overlays/actors/ovl_En_Ex_Item/z_en_ex_item.c [hunk 1]
```diff
@@ -138,8 +138,11 @@ void EnExItem_WaitForObject(EnExItem* this, GlobalContext* globalCtx) {
             case EXITEM_BOMB_BAG_BOWLING:
                 this->unk_17C = func_8002EBCC;
                 if (gSaveContext.n64ddFlag) {
-                    this->giDrawId = GetItemModelFromId(
-                        GetRandomizedItemIdFromKnownCheck(RC_MARKET_BOMBCHU_BOWLING_FIRST_PRIZE, GI_BOMB_BAG_20));
```
Comment (briaguya0) [inline soh/src/overlays/actors/ovl_En_Ex_Item/z_en_ex_item.c]: we can just use this, we don't need to do any particle stuff up here

File: soh/src/overlays/actors/ovl_Item_Etcetera/z_item_etcetera.c [hunk 2]
```diff
@@ -225,6 +225,10 @@ void ItemEtcetera_DrawThroughLens(Actor* thisx, GlobalContext* globalCtx) {
         func_8002ED80(&this->actor, globalCtx, 0);
 
         if(gSaveContext.n64ddFlag && globalCtx->sceneNum == 16) {
+            s32 randoGetItemId = GetChestGameRandoGetItemId(this->actor.room, this->giDrawId, globalCtx);
+            if (randoGetItemId >= GI_MINUET_OF_FOREST && randoGetItemId <= GI_DOUBLE_DEFENSE) {
+                EnItem00_CustomItemsParticles(&this->actor, globalCtx, randoGetItemId);
+            }
             GetItem_Draw(globalCtx, GetChestGameRandoGiDrawId(this->actor.room, this->giDrawId, globalCtx));
```
Comment (briaguya0) [inline soh/src/overlays/actors/ovl_Item_Etcetera/z_item_etcetera.c]: no need to call the chestgamerandodrawid here now that we have the getitemid, we can just
```
GetItem_Draw(globalCtx, GetItemModelFromId(randoGetItemId));
```

File: soh/src/overlays/actors/ovl_Item_Etcetera/z_item_etcetera.c [hunk 3]
```diff
@@ -242,6 +246,10 @@ void ItemEtcetera_Draw(Actor* thisx, GlobalContext* globalCtx) {
             this->giDrawId = GetItemModelFromId(GetRandomizedItemIdFromKnownCheck(RC_LH_SUN, GI_ARROW_FIRE));
```
Comment (briaguya0) [inline soh/src/overlays/actors/ovl_Item_Etcetera/z_item_etcetera.c]: looks like we forgot to test the fire arrow check

**Review by briaguya0 — CHANGES_REQUESTED** [review]
looking great! couple little things to change and one check we forgot

File: soh/src/overlays/actors/ovl_En_Ex_Item/z_en_ex_item.c (hunk 1 above)
Comment (PurpleHato) [reply soh/src/overlays/actors/ovl_En_Ex_Item/z_en_ex_item.c]: Done

File: soh/src/overlays/actors/ovl_Item_Etcetera/z_item_etcetera.c (hunk 2 above)
Comment (PurpleHato) [reply soh/src/overlays/actors/ovl_Item_Etcetera/z_item_etcetera.c]: Done

File: soh/src/overlays/actors/ovl_Item_Etcetera/z_item_etcetera.c (hunk 3 above)
Comment (PurpleHato) [reply soh/src/overlays/actors/ovl_Item_Etcetera/z_item_etcetera.c]: Done

File: soh/src/overlays/actors/ovl_Item_Etcetera/z_item_etcetera.c [hunk 4]
```diff
@@ -240,8 +244,16 @@ void ItemEtcetera_Draw(Actor* thisx, GlobalContext* globalCtx) {
     if (gSaveContext.n64ddFlag) {
         if (type == ITEM_ETC_ARROW_FIRE) {
             this->giDrawId = GetItemModelFromId(GetRandomizedItemIdFromKnownCheck(RC_LH_SUN, GI_ARROW_FIRE));
+            s32 randoGetItemId = GetRandomizedItemId(this->getItemId, this->actor.id, this->actor.params, globalCtx->sceneNum);
+            if (randoGetItemId >= GI_MINUET_OF_FOREST && randoGetItemId <= GI_DOUBLE_DEFENSE) {
+                EnItem00_CustomItemsParticles(&this->actor, globalCtx, randoGetItemId);
+            }
         } else if (type == ITEM_ETC_LETTER) {
             this->giDrawId = GetItemModelFromId(GetRandomizedItemIdFromKnownCheck(RC_LH_UNDERWATER_ITEM, GI_LETTER_RUTO));
+            s32 randoGetItemId = GetRandomizedItemId(this->getItemId, this->actor.id, this->actor.params, globalCtx->sceneNum);
```
Comment (briaguya0) [inline soh/src/overlays/actors/ovl_Item_Etcetera/z_item_etcetera.c]: didn't notice this before, but i'm surprised this is working...

it's interesting that it is, but it doesn't make sense for us to do one call to KnownCheck and one using id/params/scene

i'm thinking this whole block can be cleaned up into 
```c
    if (gSaveContext.n64ddFlag) {
        s32 randoGetItemId = GI_NONE;
        if (type == ITEM_ETC_ARROW_FIRE) {
            randoGetItemId = GetRandomizedItemIdFromKnownCheck(RC_LH_SUN, GI_ARROW_FIRE);
        } else if (type == ITEM_ETC_LETTER) {
            randoGetItemId = GetRandomizedItemIdFromKnownCheck(RC_LH_UNDERWATER_ITEM, GI_LETTER_RUTO);
        }
        
        if (randoGetItemId >= GI_MINUET_OF_FOREST && randoGetItemId <= GI_DOUBLE_DEFENSE) {
            EnItem00_CustomItemsParticles(&this->actor, globalCtx, randoGetItemId);
        }
        
        if (randoGetItemId != GI_NONE) this->giDrawId = GetItemModelFromId(randoGetItemId);
    }
```

**Review by briaguya0 — CHANGES_REQUESTED** [review]
left a comment with a way to clean up the ruto/fire arrow stuff, if that doesn't work let me know and i can merge this as is

File: soh/src/overlays/actors/ovl_Item_Etcetera/z_item_etcetera.c [hunk 5]
```diff
@@ -237,11 +241,20 @@ void ItemEtcetera_Draw(Actor* thisx, GlobalContext* globalCtx) {
     ItemEtcetera* this = (ItemEtcetera*)thisx;
     s32 type = this->actor.params & 0xFF;
 
-    if (gSaveContext.n64ddFlag) {
+        if (gSaveContext.n64ddFlag) {
```
Comment (briaguya0) [inline soh/src/overlays/actors/ovl_Item_Etcetera/z_item_etcetera.c]: super tiny nitpick but this should be a tiny bit less indented

**Review by briaguya0 — CHANGES_REQUESTED** [review]
last one i promise

**Review by briaguya0 — APPROVED** [review]
:shipit:

**Pull request merged**